### PR TITLE
New-app: add support for template files

### PR DESCRIPTION
**Pull request by csrwng**
Fixes https://github.com/openshift/origin/issues/2390

### CODE REVIEW

File: pkg/cmd/cli/cmd/newapp.go [hunk 1]
```diff
@@ -86,6 +87,7 @@ func NewCmdNewApplication(fullName string, f *clientcmd.Factory, out io.Writer)
 	cmd.Flags().VarP(&config.ImageStreams, "image", "i", "Name of an OpenShift image stream to use in the app.")
 	cmd.Flags().Var(&config.DockerImages, "docker-image", "Name of a Docker image to include in the app.")
 	cmd.Flags().Var(&config.Templates, "template", "Name of an OpenShift stored template to use in the app.")
+	cmd.Flags().VarP(&config.TemplateFiles, "file", "f", "Name of a template file to instantiate in the app.")
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/newapp.go]: "Path to a template file to use to base the application on" or similar

File: pkg/cmd/cli/cmd/newapp.go [hunk 2]
```diff
@@ -65,8 +65,9 @@ application is created.`
 
 // NewCmdNewApplication implements the OpenShift cli new-app command
 func NewCmdNewApplication(fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/newapp.go]: Add an example usage

File: pkg/generate/app/file.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package app
+
+import (
+	"os"
+)
+
+// IsFile returns true if the passed-in argument is a file in the filesystem
+func IsFile(name string) bool {
```
Comment (smarterclayton) [inline pkg/generate/app/file.go]: Don't make these public

File: pkg/generate/app/templatelookup.go [hunk 4]
```diff
@@ -45,3 +52,40 @@ func (r TemplateResolver) Resolve(value string) (*ComponentMatch, error) {
...
+// Resolve attemps to read the template file and transform it into a template object
+func (r *TemplateFileResolver) Resolve(value string) (*ComponentMatch, error) {
+	if !IsFile(value) {
+		return nil, fmt.Errorf("%s is not a file", value)
+	}
+	obj, err := resource.NewBuilder(r.Mapper, r.Typer, r.ClientMapper).
+		NamespaceParam(r.Namespace).RequireNamespace().
+		FilenameParam(value).
+		Do().
+		Object()
```
Comment (smarterclayton) [inline pkg/generate/app/templatelookup.go]: You should be using .Singular()

File: pkg/generate/app/templatelookup.go [hunk 5]
```diff
@@ -45,3 +52,40 @@ func (r TemplateResolver) Resolve(value string) (*ComponentMatch, error) {
...
+	if err != nil {
+		return nil, err
+	}
+
+	template, ok := obj.(*templateapi.Template)
+	if !ok {
+		return nil, fmt.Errorf("object in %s is not a template", value)
```
Comment (smarterclayton) [inline pkg/generate/app/templatelookup.go]: %q

**Comment by csrwng** [comment]
comments addressed

**Comment by smarterclayton** [comment]
LGTM [merge]

**Comment by csrwng** [comment]
rebased

**Pull request merged**